Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm

```diff
@@ -0,0 +1,319 @@
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+
+#include "config.h"
+#include "TextIndicatorStyleController.h"
+
+#include "TextIndicatorStyle.h"
+#include "UnifiedTextReplacementController.h"
+#include "WebPage.h"
+#include <WebCore/DocumentMarkerController.h>
+#include <WebCore/FocusController.h>
+#include <WebCore/RenderedDocumentMarker.h>
+#include <WebCore/SimpleRange.h>
+#include <WebCore/TextIndicator.h>
+#include <WebCore/TextIterator.h>
+
+namespace WebKit {
+
+TextIndicatorStyleController::TextIndicatorStyleController(WebPage& webPage)
+    : m_webPage(webPage)
+{
+}
+
+RefPtr<WebCore::Document> TextIndicatorStyleController::document() const
+{
+    if (!m_webPage) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    RefPtr corePage = m_webPage->corePage();
+    if (!corePage) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    RefPtr frame = corePage->checkedFocusController()->focusedOrMainFrame();
+    if (!frame) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    return frame->document();
+}
+
+std::optional<WebCore::SimpleRange> TextIndicatorStyleController::contextRangeForTextIndicatorStyle(const WTF::UUID& uuid) const
+{
+    auto sessionRange = m_webPage->unifiedTextReplacementController().contextRangeForSessionWithUUID(uuid);
+    if (sessionRange)
+        return sessionRange;
+
+    for (auto sessionUUID : m_activeTextIndicatorStyles.keys()) {
+        for (auto styleState : m_activeTextIndicatorStyles.get(sessionUUID)) {
+            if (styleState.styleID == uuid) {
+                auto fullSessionRange = m_webPage->unifiedTextReplacementController().contextRangeForSessionWithUUID(sessionUUID);
+                if (fullSessionRange)
+                    return WebCore::resolveCharacterRange(*fullSessionRange, styleState.range, { WebCore::TextIteratorBehavior::EmitsObjectReplacementCharactersForImages });
+            }
+        }
+    }
+    return std::nullopt;
+}
+
+void TextIndicatorStyleController::cleanUpTextStylesForSessionID(const WTF::UUID& sessionUUID)
+{
+    auto styleStates = m_activeTextIndicatorStyles.take(sessionUUID);
+    if (styleStates.isEmpty())
+        return;
+    for (auto styleState : styleStates)
+        removeTransparentMarkersForUUID(styleState.styleID);
+
+    auto unstyledRange = m_unstyledRanges.find(sessionUUID);
+    removeTransparentMarkersForUUID(unstyledRange->value->styleID);
```

Comment:
this should probably check that unstyledRange exists, and unstyledRange->value exists